lib/hanami/container.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require "dry/system/container"
+
+module Hanami
+  # Hanami private IoC
+  #
+  # @since 2.0.0
+  class Container < Dry::System::Container
+    boot(:root) do
+      start do
+        register(:root, Hanami.root)
+      end
+    end
+
+    boot(:env) do |c|
+      init do
+        use :root
+
+        begin
+          require "dotenv"
+        rescue LoadError # rubocop:disable Lint/HandleExceptions
+        end
+
+        Dotenv.load(c[:root].join(".env")) if defined?(Dotenv)
+      end
+    end
+
+    boot(:lib) do |c|
+      init do
+        use :root
+
+        $LOAD_PATH.unshift root.join("lib")
+        Hanami::Utils.require!(c[:root].join("lib", "**", "*.rb"))
+      end
+    end
```

Comment:
This shouldn't be needed. Why blindly requiring all the lib files?

Reply:
This is for convenience sake. Given under `lib/` there are repositories and entities, that are necessary for the app.

Do you think [_Auto-Registration_](https://dry-rb.org/gems/dry-system/container/) should be used here instead? Or do you suggest to let devs to manually require code?

Reply:
@jodosha sorry for replying late - *my* preference is to "require what you require", with no exceptions. Probably a good solution in Hanami would be to offer "load everything" as an option that users can turn on/off. In such case folks coming from Rails would feel more comfortable when it's turned on.

> Do you think Auto-Registration should be used here instead?

Partly yes. Auto-registration works great for all objects that are used as injectible dependencies. There are exceptions though - ie entities are not used in DI, so they shouldn't be auto-registered.